index.js

```diff
@@ -603,160 +604,133 @@ class LatticeKeyring extends EventEmitter {
-        } else {
-          return cb(null, addrs);
-        }
-      })
+  async _fetchAddresses(n=1, i=0, recursedAddrs=[]) {
+    if (!this.isUnlocked()) {
+      throw new Error('No connection to Lattice. Cannot fetch addresses.')
+    }
+    return await this.__fetchAddresses(n, i);
   }
 
-  _signTxData(txData) {
-    return new Promise((resolve, reject) => {
-      this.sdkSession.sign({ currency: 'ETH', data: txData }, (err, res) => {
-        if (err) {
-          return reject(err);
-        }
-        if (!this._syncCurrentWalletUID()) {
-          return reject('No active wallet.');
-        }
-        if (!res.tx) {
-          return reject(new Error('No transaction payload returned.'));
-        }
-        // Here we catch an edge case where the requester is asking for an EIP1559
-        // transaction but firmware is not updated to support it. We fallback to legacy.
-        res.type = txData.type;
-        if (txData.revertToLegacy) {
-          res.revertToLegacy = true;
-          res.gasPrice = txData.gasPrice;
-        }
-        // Return the signed tx
-        return resolve(res)
-      })
-    })
+  async __fetchAddresses(n=1, i=0, recursedAddrs=[]) {
+    // Determine if we need to do a recursive call here. We prefer not to
+    // because they will be much slower, but Ledger paths require it since
+    // they are non-standard.
+    if (n === 0) {
+      return recursedAddrs;
+    }
+    const shouldRecurse = this._hdPathHasInternalVarIdx();
+
+    // Make the request to get the requested address
+    const addrData = { 
+      currency: 'ETH', 
+      startPath: this._getHDPathIndices(this.hdPath, i), 
+      n: shouldRecurse ? 1 : n,
+      skipCache: true,
+    };
+    const addrs = await this.sdkSession.getAddresses(addrData);
+    // Sanity check -- if this returned 0 addresses, handle the error
+    if (addrs.length < 1) {
+      throw new Error('No addresses returned');
+    }
+    // Return the addresses we fetched *without* updating state
+    if (shouldRecurse) {
+      return await this.__fetchAddresses(n-1, i+1, recursedAddrs.concat(addrs));
+    }
+    return addrs;
   }
 
-  _getPage(increment=0) {
-    return new Promise((resolve, reject) => {
+  async _getPage(increment=0) {
+    try {
       this.page += increment;
       if (this.page < 0)
         this.page = 0;
       const start = PER_PAGE * this.page;
       // Otherwise unlock the device and fetch more addresses
-      this.unlock()
-      .then(() => {
-        return this._fetchAddresses(PER_PAGE, start)
-      })
-      .then((addrs) => {
-        const accounts = []
-        addrs.forEach((address, i) => {
-          accounts.push({
-            address,
-            balance: null,
-            index: start + i,
-          })
+      await this.unlock()
+      const addrs = await this._fetchAddresses(PER_PAGE, start)
+      const accounts = []
+      addrs.forEach((address, i) => {
```

Comment:
we could use `map()` here: 

```js
const accounts = addrs.map((address, i) => {
	  return {
	    address,
	    balance: null,
	    index: start + i,
	  };
});
```

Reply:
[52b13a0](https://github.com/GridPlus/eth-lattice-keyring/pull/33/commits/52b13a02e9654d2b9b693627bc746868dd62e980)